### ISSUE TITLE
Race condition between docker-machine and cloud-init in Digital Ocean for Debian VMs

### ISSUE THREAD

**Issue by airbornemint** [docker/machine#4675]
`apt-get` has a [well-known bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=754103), in that it uses a lock file but has no ability to wait for the lock (if it can't acquire the lock it fails). As a result, two processes attempting to run `apt-get` at the same time will, therefore, get into a fight over the lock, and only one of them will succeed in running `apt-get`.

The digitalocean driver in Docker machine uses `apt-get` to do some post-provisioning setup. It does this by waiting for SSH to become available on the provisioned VM, then SSHing into it and running several commands, some of which involve `apt-get`. 

Meanwhile, the provisioning process on Digital Ocean uses cloud-init, and SSH is enabled relatively early on in the cloud-init process. Some steps following SSH enablement in the vendor cloud-init script for Debian-based VMs on DigitalOcean involve running `apt-get`. Specifically, the vendor cloud-init uses `apt-get` to install the `do-agent` package, which is required for monitoring of VMs. 

Furthermore, DO allows users to supply additional cloud-init commands which can (and often will) make further calls to `apt-get` to install additional packages at provisioning time.

Unfortunately, because Docker machine uses "SSH is available" as a condition for running its post-provisioning (which includes running `apt-get`), because both vendor and user cloud-init run after SSH is available and potentially also call `apt-get`, and because `apt-get` is a poopy head, this leads to race conditions and various provisioning failures. I have observed all three of the following failure modes:

1. `docker-machine create` fails because it lost the lock race and its attempt to run `apt-get` resulted in an error
2. `docker-machine create` succeeds, but the provisioned droplet has crippled monitoring because vendor cloud-init lost the lock race and thus `do-agent` was never installed
3. `docker-machine create` succeeds, but the provisioned droplet is incompletely configured because user commands supplied to cloud-init lost the lock race and therefore (silently) did not run to completion

I have discussed this with Digital Ocean support staff in a support ticket (DO ticket 3540247), and their take is that this is exactly the kind of a problem that cloud-init is supposed to solve; their preferred solution is for Docker machine's DO driver to use cloud-init to do its post-provisioning setup, rather than doing it over SSH.

This problem can be reproduced by running `docker-machine create --driver digitalocean --digitalocean-monitoring --digitalocean-image=ubuntu-16-04-x64`, although the error does not occur 100% of the time and most of the time the error is subtle (monitoring ends up not being enabled, and you can look for `apt-get` errors in `/var/log/cloud-init-output.log` after cloud-init is finished to confirm `apt-get` lock errors are the cause).

You can make the problem more likely by adding `--digitalocean-userdata` with a user data file consisting of 

```
#cloud-init
runcmd:
  - "for i in {1..100}; do apt-get update; done"
```

As a partial workaround, I have been adding this to my cloud-init scripts for Debian on DO:

```
runcmd:
  - while [ ! -d /etc/systemd/system/docker.service.d ] ; do sleep 15 ; done
  - export DEBIAN_FRONTEND=noninteractive
  - apt-get update -y -qq
  - apt-get install -y -qq --no-install-recommends do-agent
```

This blocks cloud-init until docker is up, thus guaranteeing that the rest of user cloud-init script will take place after `docker-machine create` is done with its post-provisioning setup, and therefore avoiding lock race between `docker-machine` and user cloud-init. And by redoing `apt-get install do-agent` in my user script, I fix monitoring which probably didn't get installed because the vendor cloud-init script lost the race with Docker machine. 

However, this workaround is incomplete, because it is still possible for `docker-machine create` to be the loser of the race, in which case I get a hard error trying to provision the VM. 

So, to summarize:

 * Docker machine's use of SSH for post-provisioning setup creates an `apt-get` race condition on Debian-based VMs on Digital Ocean, leading to varied and random provisioning errors
 * Docker machine should instead use cloud-init for post-provisioning setup on Digital Ocean.

**Comment by aanatoly** [comment]
Hi
I have the same issue with AWS provider and solved it by setting `--engine-install-url` to point to my init script which after all installed docker

**Comment by thearrow** [comment]
Had the exact same issue. Your partial workaround worked for me and saved me some struggling - thank you!
